Feature/Authentication_Services

AbdoMagdey24 opened the pull request.

AbdoMagdey24 committed:
apply TDD on hashing password using MD5 algorithm and implement the logic

AbdoMagdey24 committed:
- make AuthenticationRepository for login and createUser
- apply TDD for this functionality

AbdoMagdey24 committed:
- refactor package name to fit  the remote

AbdoMagdey24 committed:
- add test case to cover 100% of branches

yusufnasserdev commented on build.gradle.kts:
```diff
@@ -21,6 +21,9 @@ dependencies {
     testImplementation("io.mockk:mockk:1.14.0")
+    testImplementation("org.jetbrains.kotlin:kotlin-test")
+    testImplementation("org.junit.jupiter:junit-jupiter-engine:5.7.0")
+    testImplementation("io.mockk:mockk:1.14.0")
```
Remove the duplicated dependency.

yusufnasserdev commented on src/main/kotlin/utils/Constants.kt:
```diff
@@ -0,0 +1,4 @@
+package org.damascus.utils
+
+
+val MD5_HASHING_ALGORITHM = "MD5"
```
This can be moved to where it's used as it's needed on a project-wide level.

yusufnasserdev commented on src/main/kotlin/logic/MD5HashingService.kt:
```diff
@@ -0,0 +1,26 @@
+class MD5HashingService : HashingService {
+
+    override fun <T> hashData(data: T): String {
+        val md = MessageDigest.getInstance(MD5_HASHING_ALGORITHM)
```
`md` is a bit cryptic especially as the type for it isn't specified, please use a clearer name.

yusufnasserdev commented on src/main/kotlin/logic/services/AuthenticationRepoImpl.kt:
This should be moved to the data package.

yusufnasserdev commented on src/test/kotlin/logic/services/AuthenticationRepoImplTest.kt:
```diff
@@ -0,0 +1,191 @@
+
+import User
+import org.junit.jupiter.api.Assertions.*
+
```
Remove empty line.

yusufnasserdev commented on src/main/kotlin/logic/services/AuthenticationRepository.kt:
```diff
@@ -0,0 +1,12 @@
+interface AuthenticationRepository {
+    fun login(user: User): User
+    fun createMate(requester: User, newUsername: String, rawPassword: String): Mate
+    fun createAdmin(requester: User, newUsername: String, rawPassword: String): Admin
```
This functionality doesn't exist within this system, only one `Admin` exists and it will be `pre-created`

AbdoMagdey24 replied:
I don't think so.. take another look at the system description, you will find he is mentioning (Admins) can do.
and never mention only one Admin exists and it will be pre-created

yusufnasserdev replied:
While I tend to agree with you on this, this was relayed by our mentor @AmeerAmjed and Bareq in the squad's slack discussion.

yusufnasserdev commented on src/main/kotlin/logic/services/AuthenticationRepository.kt:
```diff
@@ -0,0 +1,12 @@
+import model.Mate
+
+interface AuthenticationRepository {
+    fun login(user: User): User
```
`authenticateUser` would be a better naming for this function.

yusufnasserdev commented on src/main/kotlin/logic/services/AuthenticationRepository.kt:
```diff
@@ -0,0 +1,12 @@
+    fun login(user: User): User
+    fun createMate(requester: User, newUsername: String, rawPassword: String): Mate
+    fun createAdmin(requester: User, newUsername: String, rawPassword: String): Admin
+    fun findByUsername(username: String): User?
```
Function name doesn't relay that it finds or gets a user, should be something like `getUserbyUsername()`.

I believe a better solution here would be to depend on the User repository to fetch the user by username.

AbdoMagdey24 replied:
Yes it will be refactored when User Repo is merged as it is not my task now

yusufnasserdev commented on src/main/kotlin/logic/services/AuthenticationRepoImpl.kt:
```diff
@@ -0,0 +1,59 @@
+        return newMate
+    }
+
+    override fun createAdmin(requester: User, newUsername: String, rawPassword: String): Admin {
+        if (requester !is Admin) {
+            throw UnauthorizedActionException("create an admin")
+        }
+        if (findByUsername(newUsername) != null) {
+            throw UserAlreadyExistException(newUsername)
+        }
+        val hashedPassword = hashingService.hashData(rawPassword)
+        val newAdmin = Admin(id = UUID.randomUUID(), newUsername, hashedPassword)
+        users.add(newAdmin)
+        return newAdmin
+    }
```
Should be removed as usecase isn't provided by this system.

AbdoMagdey24 replied:
Same as I mentioned previously , I don't agree with one admin in the system so I suggest keep it now as it is

yusufnasserdev commented on src/main/kotlin/logic/services/AuthenticationRepoImpl.kt:
File formatting is poor, it lacks spaces between regions, makes it harder to read.

yusufnasserdev commented on src/main/kotlin/logic/services/AuthenticationRepoImpl.kt:
```diff
@@ -0,0 +1,59 @@
+        val hashedPassword = hashingService.hashData(rawPassword)
+        val newMate = Mate(id = UUID.randomUUID(), newUsername, hashedPassword)
+        users.add(newMate)
+        return newMate
```
Good naming 👏🏻

yusufnasserdev commented on src/test/kotlin/logic/MD5HashingServiceTest.kt:
```diff
@@ -0,0 +1,42 @@
+package logic
+
```
Remove extra empty line

yusufnasserdev commented on src/test/kotlin/logic/MD5HashingServiceTest.kt:
Tests lack the `given/when/then` comments.

yusufnasserdev commented on src/test/kotlin/logic/services/AuthenticationRepoImplTest.kt:
Tests lack the given/when/then comments.

yusufnasserdev commented on src/main/kotlin/logic/services/AuthenticationRepoImpl.kt:
```diff
@@ -0,0 +1,59 @@
+    val users: MutableList<User>
+) : AuthenticationRepository {
+
+    override fun login(user: User): User {
+        val searchedUser = findByUsername(user.username)
+            ?: throw UserNotFoundException(user.username)
+
+        if (!hashingService.verifyData(user.password, searchedUser.password)) {
+            throw InvalidPasswordException(user.password)
+        }
+        return searchedUser
+    }
```
Please re-work this function, it should accept the input login credentials `username, password`, and return a `Boolean` whether it's a successful authentication or not.

AbdoMagdey24 replied:
I agreed with parameter list should be username, password BUT it still should return the User or Null to can use it if we need

yusufnasserdev commented on src/test/kotlin/logic/services/AuthenticationRepoImplTest.kt:
```diff
@@ -0,0 +1,191 @@
+
+        val mate = authRepo.createMate(admin, username, rawPassword)
+
+        assertEquals(username, mate.username)
+        assertEquals(expectedHash, mate.password)
+        verify(exactly = 1) { hashingService.hashData(rawPassword) }
+
+        assertTrue(users.contains(mate))
```
This can be split into two test cases, one for the created mate credentials, and another for asserting that the mate is created.

yusufnasserdev commented on src/test/kotlin/logic/services/AuthenticationRepoImplTest.kt:
```diff
@@ -0,0 +1,191 @@
+    }
+
+    @Test
+    fun `createAdmin should succeed when requester is admin and username is new`() {
+        val requester = Admin(id = UUID.randomUUID(),"admin1", "hashed123")
+        val newUsername = "admin2"
+        val rawPassword = "123456"
+        val hashedPassword = "hashed456"
+
+        every { hashingService.hashData(rawPassword) } returns hashedPassword
+
+        val result = authRepo.createAdmin(requester, newUsername, rawPassword)
+
+        assertEquals(newUsername, result.username)
+        assertEquals(hashedPassword, result.password)
+        assertTrue(users.contains(result))
+    }
+
+    @Test
+    fun `createAdmin should throw UnauthorizedActionException when requester is a mate`() {
+        val requester = Mate(id = UUID.randomUUID(),"mate1", "hashedPass")
+        val newUsername = "admin2"
+        val rawPassword = "123456"
+
+        assertFailsWith<UnauthorizedActionException> {
+            authRepo.createAdmin(requester, newUsername, rawPassword)
+        }
+    }
+
+
+    @Test
+    fun `createAdmin should throw UserAlreadyExistException when username already exists`() {
+        val requester = Admin(id = UUID.randomUUID(),"admin1", "hashed123")
+        val existingUsername = "admin2"
+        users.add(Admin(id = UUID.randomUUID(),existingUsername, "hashedExist"))
+
+        assertFailsWith<UserAlreadyExistException> {
+            authRepo.createAdmin(requester, existingUsername, "any")
+        }
+    }
```
Remove as not needed for this system usecases.

yusufnasserdev reviewed (changes requested):
Please merge with develop, resolve the requested changes, and re-request the review.

Overall great work, keep it up!

AbdoMagdey24 committed:
modify login function to take username, password instead of User
add given/when/then comments to the test cases
separate 2 test cases related to mate creation
clean variables name
move constant to its own only usage place
remove duplicated dependancies from gradle

AbdoMagdey24 committed:
add test function to cover uncovered unused unrelated task classes to pass from ci/cd

AbdoMagdey24 committed:
Merge branch 'refs/heads/develop' into feature/hash-password

# Conflicts:
#	src/main/kotlin/model/Admin.kt
#	src/main/kotlin/model/Mate.kt
#	src/main/kotlin/model/User.kt

AbdoMagdey24 committed:
use the models from logic package after merge remote
delete unneeded coverage test function after merge remote

AbdoMagdey24 committed:
Merge branch 'refs/heads/develop' into feature/hash-password

AbdoMagdey24 committed:
attach login service with csv data source

yusufnasserdev commented on src/test/kotlin/data/AuthenticationRepoImplTest.kt:
```diff
@@ -0,0 +1,201 @@
+        verify(exactly = 1) { hashingService.hashData(rawPassword) }
+    }
+
+    @Test
+    fun `createMate should add mate to users list`() {
+        // Given
+        val admin = Admin(UUID.randomUUID(), "admin1", "hashed123", Role.ADMIN)
+        val username = "newMate"
+        val rawPassword = "password"
+        val expectedHash = "hashedPassword"
+
+        val capturedMate = slot<Mate>()
+        var called = false
+
+        every { usersDataSource.read() } answers {
+            if (!called) {
+                called = true
+                emptyList()
+            } else {
+                listOf(capturedMate.captured)
+            }
+        }
+        every { hashingService.hashData(rawPassword) } returns expectedHash
+        every { usersDataSource.write(capture(capturedMate)) } returns Unit
+
+        // When
+        val mate = authRepo.createMate(admin, username, rawPassword)
+
+        // Then
+        assertTrue(usersDataSource.read().contains(mate))
+    }
```
This can be ignored for now with `@Ignore` as it's considered an integration test.

yusufnasserdev commented on src/test/kotlin/data/AuthenticationRepoImplTest.kt:
```diff
@@ -0,0 +1,201 @@
+        val exception = assertFailsWith<UserNotFoundException> {
+            authRepo.login(username, password)
+        }
+        assertEquals("User 'unknown' not found", exception.message)
```
Please avoid having multiple assertions per test when possible.

yusufnasserdev reviewed (changes requested):
Great work!

AbdoMagdey24 committed:
avoid multiple assert
ignore integration test function

yusufnasserdev reviewed (approved):
Good work!

AmeerAmjed commented on src/main/kotlin/logic/exception/CustomException.kt:
```diff
@@ -0,0 +1,9 @@
+
+class UserNotFoundException(username: String) : Exception("User '$username' not found")
+
+class InvalidPasswordException(password: String) : Exception("Invalid password '$password'")
```
Don't handle it as a specific case for a password. Use it as a generic one, like InvalidCredentialException

AbdoMagdey24 replied:
Done

AbdoMagdey24 committed:
modify invalid password exception to be invalid credentials

AbdoMagdey24 committed:
delete unneeded imports
